StripePaymentSheet/StripePaymentSheet/Source/Internal/API Bindings/Link/PaymentDetails.swift

```diff
@@ -18,98 +18,62 @@ typealias ConsumerSessionWithPaymentDetails = (session: ConsumerSession, payment
-                                      allResponseFields: dict) as? Self
+    private enum CodingKeys: String, CodingKey {
+        case stripeID = "id"
+        case isDefault
     }
 
-    class ShareResponse: NSObject, STPAPIResponseDecodable {
-        let paymentMethodID: String
-        let allResponseFields: [AnyHashable: Any]
-
-        init(paymentMethodID: String,
-             allResponseFields: [AnyHashable: Any]) {
-            self.paymentMethodID = paymentMethodID
-            self.allResponseFields = allResponseFields
-            super.init()
-        }
-
-        static func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
-            guard let dict = response,
-                  let id = dict["payment_method"] as? String else {
-                return nil
-            }
-            return ShareResponse(paymentMethodID: id,
-                                 allResponseFields: dict) as? Self
-        }
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.stripeID = try container.decode(String.self, forKey: .stripeID)
+        // The payment details are included in the dictionary, so we pass the whole dict to Details
+        self.details = try decoder.singleValueContainer().decode(Details.self)
+        self.isDefault = try container.decode(Bool.self, forKey: .isDefault)
     }
-
 }
 
 // MARK: - Details
 /// :nodoc:
 extension ConsumerPaymentDetails {
-    enum DetailsType: String, CaseIterable {
-        case card
-        case bankAccount = "bank_account"
+    enum DetailsType: String, CaseIterable, Codable {
+        case card = "CARD"
+        case bankAccount = "BANK_ACCOUNT"
     }
 
-    enum Details {
-
+    enum Details: Decodable {
         case card(card: Card)
         case bankAccount(bankAccount: BankAccount)
 
-        init?(_ response: [AnyHashable: Any]) {
-            guard let typeString = response["type"] as? String else {
-                return nil
-            }
-
-            switch typeString.lowercased() {
-            case "card":
-                if let card = Card.decodedObject(fromAPIResponse: response) {
-                    self = .card(card: card)
-                } else {
-                    return nil
-                }
-            case "bank_account":
-                if let bankAccount = BankAccount.decodedObject(fromAPIResponse: response) {
-                    self = .bankAccount(bankAccount: bankAccount)
-                } else {
-                    return nil
-                }
-            default:
-                return nil
+        private enum CodingKeys: String, CodingKey {
+            case type
+            case card = "cardDetails"
+            case bankAccount = "bankAccountDetails"
+        }
+
+        // Our JSON structure doesn't align with Swift's expected structure for enums with associated values, so we do custom decoding.
+        init(from decoder: Decoder) throws {
```

Comment:
I've seen this pattern when decoding a few other `Oneof`-shaped objects in our API. I wonder if it'd be worth making this more generic...

Reply:
Seems useful! I'll file a ticket.